Write DB Updates in Bulk; Summarize Internal Hosts After Analysis; Documentation Updates

Zalgo2462 opened the pull request:
This write up is a work in progress. I intend to list the commits that make up this PR and given an explanation behind each one. 


This PR arose from investigating the needed changes to change the max duration score for each internal host in the `host` collection to the max **total** duration score. While investigating this, I discovered two simple ways to speed up this analysis. I'd like to pull in these changes before switching over the duration score.

- Use bulk writes to send updates to MongoDB
  - This massively reduces the time spent in kernel syscalls waiting for the socket writes to finish out
- Find the highest scoring `uconn` entries for each internal host **after** running the `uconn` analysis instead of updating the `host` entries **during** the `uconn` analysis. 
  - This changes the runtime of the max scoring analysis 
    - from `O(|internal hosts| * |external hosts|)` 
    - to `O(|internal hosts| * log(|internal hosts|) + |internal hosts| * log(|external hosts|))`

I've implemented these two changes where they were applicable. These changes need to be tested on rolling datasets and the output needs to be compared to that of the current version of RITA. I have done substantial testing with dnscat and empire datasets, manually importing them in chunks, to make sure there aren't any issues. However, we should obtain larger datasets to test with. 

On small datasets, I have seen a 25% reduction in runtime. I expect the reductions to be larger for larger datasets.

While implementing these changes, I noticed that most of the comments in the code had grown useless. So, I decided to take the time to edit some of the comments as well. Additionally, I have taken the time to document the inputs and outputs of each RITA package by placing a Readme.md in each package's directory.  We may want to break the readme updates out into their own PR. 

Since this PR is large, I've tried to group the changes into meaningful commits. It might be easier to comment on the individual commits than the PR itself. Additionally, we can split up this PR by package if need be. 

## Commits in chronological order

#### Update mgo version to get access to better bulk API handlers:
- ba00b9a Update mgo dependencies to the same version used in other activecm projects
  - The bulk APIs in the old version require you to manually break up your batches into groups of 1000 records.

#### Updates to `pkg/host`:
- 1515798 Move exploded dns summary in host package to a 2nd pass
  - Separated the MongoDB updates in the analyzer into their own functions
  - Moved the `max_dns` analysis to a separate summary phase after the analysis phase
  - Changed the writer to use bulk upserts
- 674e4f3 Only compute host dns summary for local hosts
  - Currently RITA will compute the host `max_dns` analysis for external hosts if the data is available. This commit disabled this analysis for external hosts since we don't need it.
- a4a0eda fix host pkg comments

#### Updates to `pkg/uconn`:
- 33ecfd2 Run host update of maxdur after uconn pass
  - Moved `max_duration` analysis to a a separate summary phase after the analysis phase
  - Changed the writer to use bulk upserts
  - Accidentally included profiling code and files ( removed in 11ab215 )
- 768d975 Clean up uconn analyzer
  - Separated the MongoDB updates in the analyzer into their own functions
- f7d3290 fix uconn pkg comments
- 10f6d89 remove profiling code in uconn package
- 11ab215 Remove pprof files from git index

#### Updates to `pkg/uconnproxy`:
- 5ca7cc0 clean up uconn proxy analyzer and comments
  - Removed duplicate code in the analyzer
  - Broke out the main MongoDB update in the analyzer out into its own function
  - Cleaned up comments in `mongodb.go`
  - Fixed an error in `CreateIndexes()` which resulted in sending a malformed request for MongoDB to index the `dat.count` field. As a result, the `dat.count` field in the `uconnproxy` collection is not properly indexed by RITA.
  - Changed the writer to use bulk upserts
  
#### Updates to `pkg/explodeddns`:
- b9e9eef bulk writer for exploded dns; comment cleanup

#### Updates to `pkg/hostname`:
- 39ffbed cleanup hostname pkg; bulk writer for hostname pkg
  - Separated the MongoDB updates in the analyzer into their own functions
  - Cleaned up comments in `mongodb.go`
  - Changed the writer to use bulk upserts
  
#### Updates to `pkg/beacon`:
- a261693 Move max beacon into a summarize phase; clean up beacon analysis; remove icerts from beacon package
  - Changed the writer to use bulk upserts
    - This package uses a more complex callback based bulk writer since it needs to handle running both removes and upserts from two different collections
  - Removes the invalid certificate analysis from the beacons package (moved to uconns  in the summarizer in 87dee2a)
  - Moved `mbdst` analysis to a a separate summary phase after the analysis phase

#### Updates to `pkg/uconn` (dependent on beacon changes):
- 87dee2a Move icerts analysis from beacon package to uconn package
- ee3a08f uconn updates
  - Fixed an error in `CreateIndexes()` which resulted in sending a malformed request for MongoDB to index the `dat.count` field. As a result, the `dat.count` field in the `uconnproxy` collection is not properly indexed by RITA.
   - Fixed omments and linter issues
   
#### Updates to `pkg/beaconfqdn`:
- 1b021da Move max fqdn beacon analysis into separate summary step
  - Moved `mbfqdn` analysis to a separate summary phase after the analysis phase
  - Changed the writer to use bulk upserts
  
#### Updates to `pkg/beaconproxy`:
- e584710 Move max proxy beacon analysis into separate summary step
  - Changed the writer to use bulk upserts
    - This package uses a more complex callback based bulk writer since it needs to handle running both removes and upserts from two different collections
  - Moved `mbproxy` analysis to a a separate summary phase after the analysis phase 
- 65c9b28 remove buggy strobeFQDN field from the beaconProxy collection; this is tracked in the uconnProxy collection
- 67c616b remove tslist field from beaconProxy to match other collections; this info is stored in the uconnProxy collection

#### Updates to `pkg/useragent`:
- 2e73b0a Move useragent host rare signature analysis to summary phase
    - Moved `rsigc` analysis to a a separate summary phase after the analysis phase 
      - I believe there is a currently a bug in the rsigc analysis where we are missing some records in the  current version
  - Changed the writer to use bulk upserts
    
#### Updates to `pkg/certificate`:
- 544fe2b use bulk writer for certificate package

#### Updates to `pkg/blacklist`:
- 9a03eb3 Add bulk writer and error reporting to blacklist package

#### Preventing some duplicate summary entries during rolling imports:
These commits changed the summarizer calling code to loop through internal hosts in the current hostMap instead of calling out to MongoDB to find internal hosts in this chunk. This prevents creating duplicate entries when RITA has to run multiple imports into the same chunk for resource reasons.

- 9fb546a Use fsimporter hostMap instead of mgo in order to find local hosts to summarize in beacons
- 38af022 Use fsimporter hostMap instead of mgo in order to find local hosts to summarize in proxybeacons
- 9aaa738 Use fsimporter hostMap instead of mgo in order to find local hosts to summarize in uconns

#### Package readme updates:
- 1a408d7 Add host package Readme.md
- dd863e7 uconn pkg readme
- ad18473 uconnproxy readme
- 22ceb1d exploded dns readme
- 0646835 hostname pkg readme
- 8ada832 beacon pkg readme
- c1d39d3 beaconfqdn readme
- cc5ce90 beaconproxy pkg readme
- e9865ff Useragent pkg readme:
- 07c185d certificate pkg readme
- 6cb7216 threat intel peers pkg readme

#### CI fixes:
- 907f85e Fix broken link for golangci-lint in testing dockerfile

caffeinatedpixel commented on pkg/beaconfqdn/writer.go:
```diff
@@ -44,44 +45,38 @@ func (w *writer) close() {
-				if err != nil ||
-					((info.Updated == 0) && (info.UpsertedId == nil)) {
+			if count >= 1000 {
+				info, err := bulk.Run()
```
The bulk upserts that are run here are sometimes causing Mongo to error because part or all of the upsert statement provided is over 16MB in size. When enough of these errors occur, it causes the entire MongoDB server to crash.

Attached are the RITA error logs. 
[error-bulk.txt](https://github.com/activecm/rita/files/8969150/error-bulk.txt)

Zalgo2462 replied:
Good catch! I'll cut the limit in half with a comment explaining why. We could probably get a way with 900, but I think 500 gets us a little more head room.

Zalgo2462 replied:
Pushed a fix

caffeinatedpixel reviewed (changes requested):
Submitting a partial review due to the issue within the beaconFQDN package. Will continue to review.

Zalgo2462 commented on pkg/beacon/mongodb.go:
```diff
@@ -129,4 +131,53 @@ func (r *repo) Upsert(uconnMap map[string]*uconn.Input, minTimestamp, maxTimesta
+	p = mpb.New(mpb.WithWidth(20))
+	bar = p.AddBar(int64(len(localHosts)),
+		mpb.PrependDecorators(
+			decor.Name("\t[-] Beacon Analysis (2/2):", decor.WC{W: 30, C: decor.DidentRight}),
```
I'll change all of these together at the end. Still trying to decide on the right naming.

Zalgo2462 commented on pkg/beaconfqdn/Readme.md:
```diff
@@ -0,0 +1,383 @@
+### Highest Scoring FQDN Beacon Summary
+Inputs: 
+- `ParseResults.HostMap` created by `FSImporter`
+    - Field: `IsLocal`
```
Fixed

Zalgo2462 commented on pkg/beacon/Readme.md:
```diff
@@ -0,0 +1,261 @@
+
+Inputs: 
+- `ParseResults.HostMap` created by `FSImporter`
+    - Field: `IsLocal`
```
Fixed

Zalgo2462 commented on pkg/beacon/Readme.md:
```diff
@@ -0,0 +1,261 @@
+    - Field: `score`
+        - Type: float64
+
+Outputs:
```
Fixed

caffeinatedpixel commented on pkg/beaconproxy/writer.go:
```diff
@@ -5,105 +5,202 @@ import (
 		w.writeWg.Done()
 	}()
 }
+
```
Remove large commented block

Zalgo2462 replied:
Fixed

caffeinatedpixel reviewed (changes requested):
The beaconFQDN and host packages are still having issues with the 16MB limit. I think that the `dat.sizes` & `dat.count` fields are the culprit for the beaconFQDN package. Some fqdn beacons can have very large arrays for these two fields. 
The explodedDns array is probably the culprit for the host package, but I don't have any data to prove that.

Zalgo2462 commented on pkg/beaconproxy/Readme.md:
```diff
@@ -0,0 +1,208 @@
+        - Type: float64
+
+Outputs:
+    - Array Field: `dat`
```
Fixed

Zalgo2462 commented:
> The beaconFQDN and host packages are still having issues with the 16MB limit. I think that the `dat.sizes` & `dat.count` fields are the culprit for the beaconFQDN package. Some fqdn beacons can have very large arrays for these two fields. The explodedDns array is probably the culprit for the host package, but I don't have any data to prove that.

I've temporarily limited the bulk buffer size to just one document. Please test whether we still hit the 16MB limit. 
If this fixes it, I will try to come up with a buffer limit based on the size of the documents.

Do we know if master-branch RITA does not throw the same error?

caffeinatedpixel commented:
> I've temporarily limited the bulk buffer size to just one document. Please test whether we still hit the 16MB limit. If this fixes it, I will try to come up with a buffer limit based on the size of the documents.
> 
The test is still running and probably won't be done for another hour or two.
> Do we know if master-branch RITA does not throw the same error?
>
The master branch does not appear to throw the same error. There are fqdns that trigger the error that aren't in the dataset that the master branch produces. I wonder if the fqdn gatherer is picking up fqdns with larger data that the master version doesn't pick up, and when bulked together, hit the 16mb maximum.

Zalgo2462 commented:
Since the `host` package had issues with a single document write buffer, I think the bug is probably occurring in an earlier stage. Do you happen to know if the error occurred during the analysis or summary phase? (`[-] Host Analysis (1/2):` vs `[-] Host Analysis (2/2):`)

caffeinatedpixel commented:
It occurs during the analysis phase. The host that causes the error has an obscene amount of fqdns in the `exploded_dns` array, none of them being duplicates. The same host in the master branch import only has a tiny fraction of the amount of fqdns in the `exploded_dns` array in this pr's import. I couldn't find any matching errors for the host package errors with the master import.

caffeinatedpixel commented:
I found the cause of the `exploded_dns` array count discrepancy, so we can ignore that issue. The host that causes the write error will still be inserted, but will not have an `exploded_dns` array, and subsequently no results for the `max_dns` array, leaving the host blind to c2 over dns.  
We can leave the host package as is, as I assume the master branch would error out if a host had over 16MB of exploded dns data. Other options are to batch the `exploded_dns` inserts, limit the amount of entries inserted, or write a message to the console recommending that the user adds some domains to the `NeverIncludeDomain` section of the config. 

I'm still waiting on the rerun test to complete. If it turns out that running single inserts creates no issues for the beaconFQDN package, I'd recommend putting beacons with large `ds.sizes` & `ds.counts` arrays into a separate queue to be inserted one by one.

Zalgo2462 commented:
Good to hear about the `exploded_dns` discrepancy. We may want to make a separate ticket to address this in the future.

I have a patch written that will flush the write buffer whenever we would a new entry to the buffer which would break over the 16 MB limit. I'd like to try this patch out if the `beaconFQDN` package runs successfully when limited to 1 document per write.

It serializes the MongoDB selector and query objects to the MongoDB wire format and keeps track of the size of the objects in the write buffer. I'm a bit worried all the extra serialization work might slow down the the process though. 

I think using the length of `ds.sizes` aught to be a good estimator as to whether a document is large enough to burst over the 16 MB limit. So, if the approach I have written up doesn't work out, I'll take a whack at implementing the separate queue you're suggesting.

caffeinatedpixel commented:
The test finally finished. The beaconFQDN package has no issues running one update at a time.

Zalgo2462 commented:
I've pushed up a patch for the beaconFQDN package which tracks the size of the serialized MongoDB documents and makes sure to send the buffer over before the buffer hits 16MB. This commit also restores the buffer in the host package to 500 docs.

https://github.com/activecm/rita/pull/737/commits/7750de1d8d3db467f7bc4e0418c09752a9e8e5d8

caffeinatedpixel reviewed (commented):
https://github.com/activecm/rita/commit/7750de1d8d3db467f7bc4e0418c09752a9e8e5d8 fixed the issue with the beaconFQDN package. 

Verified that there are no `rsig` and `icdst` entries in the rolling dataset created by master that do not exist in the rolling dataset created by this PR. There are quite a few `rsig` and `icdst` entries in this pr's rolling dataset that do not exist in the master dataset. All of these entries are for internal hosts and for `icdst`, map to external destinations. Is this to be expected? If it is, then I'll approve the PR.

I noticed about a 40% speed improvement on imports with this PR.

Zalgo2462 commented:
In the master branch, `icdst` records are only created if the unique connection pair has enough connections to be considered a beacon and the destination presented an invalid certificate. In this PR, the beacon connection threshold is not considered. 

In my testing, I found that all of the `icdst` entries that were included in the output from the PR which were not included in the output from the master branch also showed up in the `cert` collection. This leads me to believe that either all of these new entries can be explained with the removal of the connection threshold or that there may be a bug in the master branch. 

I need to do some testing on `rsig` tonight. I'm not sure what's going on there to be honest.

Zalgo2462 commented:
I have confirmed that during rolling imports the code in the PR creates `rsig` entries that should not exist in the `host` collection. I will work on tracking down the source of the bug tonight.

Zalgo2462 commented:
I have tracked down the source of the extra `rsig` entries. The query to count the number of `orig_ips` associated with a given useragent from the previous chunks was being run against the `uconn` collection instead of the `useragent` collection. :facepalm: 

After fixing this issue, I have found that the `rsig` entries match the master branch entries.

Zalgo2462 commented:
Alrighty, I've fixed up the `rsig` issue, and changed the status text for the different analysis phases. I think we're good to go :+1: 

I'm pretty well convinced that the extra `icdst` entries are created since this PR doesn't check the connection threshold when creating `icdst` entries. I can add this threshold in if we need it, but I don't see why we'd want to tie invalid certificate markers to a minimum connection count.
